include/pmacc/math/complex/Bessel.tpp

```diff
@@ -0,0 +1,271 @@
+ *  software is offered freely and without restriction -- although
+ *  in most cases the files should be considered as works in progress
+ *  (alpha or beta level). Source code is also included for some
+ *  applications."
+ *
+ *  Code history:
+ *  1/03 -- Added C/C++ source files for real and complex gamma function
+ *  and psi function. Also added individual C/C++ files  for Bessel and
+ *  modified Bessel functions of 1st and 2nd kinds for real and complex
+ *  arguments. Updated Butterworth and Bessel filter tables with files of
+ *   extended parameters including polynomials, poles and component values.
+ *  6/04 -- Revised bessel.zip to correct errors in complex Bessel functions.
+ *
+ *  Further (re-)implementation of this code has been done by FZ Juelich.
+ *  URL: http://apps.jcns.fz-juelich.de/redmine/issues/569#change-2056
+ *  Above URL also includes a accuracy test report of this code against
+ *  SLATEC, MAPLE and MATHEMATICA.
+ */
+
+#pragma once
+
+#include <pmacc/types.hpp>
+#include "pmacc/algorithms/math.hpp"
+#include "pmacc/algorithms/TypeCast.hpp"
+#include "pmacc/math/Complex.hpp"
+
+#include <cmath>
+
+
+namespace pmacc
+{
+namespace algorithms
+{
+namespace math
+{
+    namespace pmMath = pmacc::algorithms::math;
+
+    template<
+        typename T_FloatType,
+        typename T_TableA,
+        typename T_TableB,
+        typename T_TableA1,
+        typename T_TableB1
+    >
+    struct Cbesselj0Base
+    {
+        typedef pmacc::math::Complex< T_FloatType > result;
+        typedef pmacc::math::Complex< T_FloatType > complex_T;
+        typedef T_FloatType float_T;
+        T_FloatType const eps = T_FloatType( 1.0e-15 );
+        T_FloatType const el = T_FloatType( 0.5772156649015329 );
+
+        complex_T const cii = complex_T( 0, 1 );
+        complex_T const cone = complex_T( 1, 0 );
+        complex_T const czero = complex_T( 0, 0 );
+
+        HDINLINE result operator( )( complex_T const z )
+        {
+            T_TableA a;
+            T_TableB b;
+            T_TableA1 a1;
+            T_TableB1 b1;
+            result cj0;
+
+            float_T const a0 = pmMath::abs( z );
+            float_T const z2 = z * z;
+            complex_T z1 = z;
+            if ( a0 == float_T( 0.0 ) )
+            {
+                cj0 = cone;
+                return cj0;
+            }
+            if ( z.get_real() < float_T( 0.0 ) ) z1 = -z;
+            if ( a0 <= float_T( 12.0 ) )
+            {
+                cj0 = cone;
+                complex_T cr = cone;
+                for ( uint32_t k = 1 ; k <= 40 ; k++ )
+                {
+                    cr *= float_T( -0.25 ) * z2 / float_T( k * k );
+                    cj0 += cr;
+                    if ( pmMath::abs( cr ) < pmMath::abs( cj0 ) * eps ) break;
+                }
+            }
+            else {
+                uint32_t kz;
+                if ( a0 >= float_T( 50.0 ) ) kz = 8;      // can be changed to 10
```

Comment:
Then we change the accuracy of the result by taking into account more or less terms of the series.